python/ql/test/query-tests/Expressions/general/expressions_test.py

```diff
@@ -242,3 +242,52 @@ def func():
 
 def mpt_arg(d=MappingProxyType({})):
     return 1 in d
+
+
+
+
+
+
+
+
+
+
+
+
+#### TruncatedDivision.ql
+
+# NOTE: The following test case will only work under Python 2.
+
+# Truncated division occurs when two integers are divided. This causes the
+# fractional part, if there is one, to be discared. So for example, `2 / 3` will
+# evaluate to `0` instead of `0.666...`.
+
+def truncated_division():
+
+  def average(l):
+    return sum(l) / len(l)
+
+
+
+  ## Negative Cases
+
+  # This case is good, and is a minimal obvious case that should be good. It
+  # SHOULD NOT be found by the query.
+  print(3.0 / 2.0)
+
+  # This case is good, because the sum is `3.0`, which is a float, and will not
+  # truncate. This case SHOULD NOT be found by the query.
+  print(average([1.0, 2.0]))
+
+
+
+  ## Positive Cases
+
+  # This case is bad, and is a minimal obvious case that should be bad. It
+  # SHOULD be found by the query.
+  print(3 / 2)
+
+  # This case is bad, because the sum is `3`, which is an integer, and will
+  # truncate when divided by the length `2`. This case SHOULD be found by the
+  # query.
+  print(average([1,2]))
```

Comment:
Despite us using it as an example, I don't actually expect that we'll be able to track things accurately enough through the call to `average` in order to figure out that there's truncation going on here.

Reply:
Shall I remove these cases then? They currently contribute nothing to the results so the entries in the test file are for the future, really.

Also I can explore what sort functions we might want to support with some amount of value reasoning. For instance, length really ought to be known to produce an integer, sum ought to be known to produce some kind of number, etc.

Reply:
I say leave the test cases in, just in case we improve the analysis in the future. (It's always helpful to document the limits of these things.) The comment should probably be adjusted accordingly, though.

As for `len` and `sum` I agree that the former should return an integer (but really it could return anything, if the object overrides `__len__`). As for `sum`, it should return the same type as the items of the iterable it receives as an argument, though again this depends on what special methods the underlying type has.